mods/classic-maximized-windows-fix.wh.cpp

```diff
@@ -242,90 +228,98 @@ bool InitCriticalSections()
 {
-    HRESULT hr = S_OK;
-    DWORD currentThreadId = GetCurrentThreadId();
-    Wh_Log(L"Entering critical section in thread #%d", currentThreadId);
-    EnterCriticalSection(&g_lockForAffectedThreads);
+    BOOL result = ThemePostWndProc_orig(hWnd, uMsg, wParam, lParam, plResult, ppvParam);
 
-    if (!g_affectedThreads.contains(currentThreadId))
-    {
-        HHOOK hHook = SetWindowsHookExW(
-            WH_CALLWNDPROC,
-            (HOOKPROC)OnCallWndProc,
-            NULL,
-            currentThreadId
-        );
-
-        if (hHook)
-        {
-            Wh_Log(
-                L"Installed Windows Hook #%d (%p) for thread #%d (%p)",
-                hHook, hHook,
-                currentThreadId, currentThreadId
-            );
-            g_affectedThreads.insert(
-                std::make_pair(currentThreadId, hHook)
-            );
-        }
-        else
-        {
-            hr = E_FAIL;
-        }
-    }
-    Wh_Log(L"Leaving critical section in thread #%d", currentThreadId);
-    LeaveCriticalSection(&g_lockForAffectedThreads);
+    HandleMaxForWindow(hWnd, uMsg);
 
-    return hr;
+    return result;
 }
 
+struct CMWF_SYMBOL_HOOK {
+    std::vector<std::wstring> symbols;
+    void** pOriginalFunction;
+    void* hookFunction = nullptr;
+    bool optional = false;
+    void **pSharedMemoryCache;
+};
+
 /*
- * UninstallHooks: Uninstall all Windows Hooks for every thread.
+ * CmwfHookSymbols: A custom hook wrapper which allows storing symbol hook results in the
+ *                  module's shared memory for faster access.
  */
-void UninstallHooks()
+bool CmwfHookSymbols(
+        HMODULE module,
+        const CMWF_SYMBOL_HOOK *symbolHooks,
+        size_t symbolHooksCount
+)
 {
-    DWORD currentThreadId = GetCurrentThreadId();
-    Wh_Log(L"Entering critical section in thread #%d", currentThreadId);
-    EnterCriticalSection(&g_lockForAffectedThreads);
-
-    int installedHooks = g_affectedThreads.size();
-
-    for (auto &it : g_affectedThreads)
+    for (size_t i = 0; i < symbolHooksCount; i++)
     {
-        HHOOK &hHook = it.second;
-        BOOL success = UnhookWindowsHookEx(hHook);
-
-        if (success)
-            installedHooks--;
-    }
+        void *address = nullptr;
+        if (symbolHooks[i].pSharedMemoryCache && *(symbolHooks[i].pSharedMemoryCache) != NULL)
+        {
+            address = *(symbolHooks[i].pSharedMemoryCache);
+            Wh_Log(
+                L"CmwfHookSymbols: Hooking symbol %.*s from in-memory cache.",
+                symbolHooks[i].symbols[0].length(),
+                symbolHooks[i].symbols[0].data()
+            );
+        }
+        else
+        {
+            address = nullptr;
+            WindhawkUtils::SYMBOL_HOOK proxyHook = {
+                symbolHooks[i].symbols,
+                &address,
+                NULL,
+                symbolHooks[i].optional
+            };
+
+            if (!WindhawkUtils::HookSymbols(module, &proxyHook, 1))
```

Comment:
I don't see much value in this wrapper. Is it to have an in-memory cache instead of a registry cache? Did you measure the difference?

Also, this mod only hooks one symbol, but generally this is going to be very slow, since this wrapper handles each symbol separately, and so the symbols will be enumerated again for each symbol, and enumerating symbols is usually the slow part in symbol loading.

Reply:
Yes, that is the goal. I've noticed that application startup can still be quite slow, even with the registry cache, so I wanted to try something else out to make it nicer to use for hooking all running applications. It seems that the slowness of reading from the registry is dependent on the application in particular, but there are some for which it can be particularly noticeable (in my experience, Git and some old versions of Notepad).

I will improve the wrapper for future usage later, but I think it should be fine here since it's only used for the one symbol.

Reply:
I don't mind merging it, but I doubt that the single registry read causes the slowness, Windhawk reads other values from the registry, the cache value really shouldn't matter. That's unlike the first run when the cache isn't available, for which symbols have to be downloaded and enumerated, but your wrapper doesn't improve this case.

Reply:
I elaborated on the possible inefficiency of the wrapper in the latest commit.